arrows/vxl/color_commonality_filter.cxx

```diff
@@ -0,0 +1,561 @@
+filter_color_image( vil_image_view< InputType > const& input,
+                    vil_image_view< OutputType >& output,
+                    std::vector< unsigned >& histogram,
+                    color_commonality_filter_settings const& options )
+{
+  assert( input.ni() == output.ni() && input.nj() == output.nj() );
+  assert( is_power_of_two( options.resolution_per_channel ) );
+
+  if( input.ni() == 0 || input.nj() == 0 )
+  {
+    return;
+  }
+
+  // Configure output scaling settings based on the output type and user
+  // settings
+  InputType const input_type_max = std::numeric_limits< InputType >::max();
+  OutputType const output_type_max = std::numeric_limits< OutputType >::max();
+  auto const histogram_threshold = static_cast< unsigned >( output_type_max );
+  unsigned histogram_scale_factor = options.output_scale_factor;
+
+  // Use type default options if no scale factor specified
+  if( histogram_scale_factor == 0 )
+  {
+    histogram_scale_factor = histogram_threshold;
+  }
+
+  // Populate histogram steps for each channel of the hist
+  std::vector< std::ptrdiff_t > histsteps( input.nplanes() );
+  histsteps[ 0 ] = 1;
+
+  for( unsigned p = 1; p < input.nplanes(); ++p )
+  {
+    histsteps[ p ] = histsteps[ p - 1 ] * options.resolution_per_channel;
+  }
+
+  // Fill in histogram of the input image
+  unsigned const bitshift =
+    integer_log2( ( static_cast< unsigned >( input_type_max ) + 1 ) /
+                  options.resolution_per_channel );
+
+  unsigned* hist_top_left = &histogram[ 0 ];
+
+  populate_image_histogram( input, hist_top_left, bitshift, &histsteps[ 0 ] );
+
+  // Normalize histogram to the output types range
+  unsigned sum = 0;
+
+  for( unsigned i = 0; i < histogram.size(); ++i )
+  {
+    sum += histogram[ i ];
+  }
+
+  // Fill in color commonality image from the compiled histogram
+  for( unsigned i = 0; i < histogram.size(); ++i )
+  {
+    unsigned value = ( histogram_scale_factor * histogram[ i ] ) / sum;
+    histogram[ i ] =
+      ( value > histogram_threshold ? histogram_threshold : value );
+  }
+
+  auto const ni = input.ni();
+  auto const nj = input.nj();
+  auto const np = input.nplanes();
+  auto const istep = input.istep();
+  auto const jstep = input.jstep();
+  auto const pstep = input.planestep();
+
+  InputType const* row = input.top_left_ptr();
+  for( unsigned j = 0; j < nj; ++j, row += jstep )
+  {
+    InputType const* pixel = row;
+    for( unsigned i = 0; i < ni; ++i, pixel += istep )
+    {
+      InputType const* plane = pixel;
+      std::ptrdiff_t step = 0;
+      for( unsigned p = 0; p < np; ++p, plane += pstep )
+      {
+        step += histsteps[ p ] * ( *plane >> bitshift );
+      }
+      output( i, j ) = static_cast< OutputType >( *( hist_top_left + step ) );
+    }
+  }
+}
+
+// ----------------------------------------------------------------------------
+/// Create an output image indicating the relative commonality of each
```

Comment:
I think this documentation either doesn't belong here, or shouldn't be doxygen. It could also stand to be rewrapped, and "occurring" is misspelled.